### PR TITLE
Extraction of coverage data from UDFs

### PR DESCRIPTION
**Pull request by jakobbraun**

### CODE REVIEW

File: src/test/java/com/exasol/LogProxy.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+        private Logger(final Socket socket) {
+            this.socket = socket;
+        }
+
+        @Override
+        public void run() {
+            try {
+                System.out.println("connected logger");
+                final BufferedReader reader = new BufferedReader(new InputStreamReader(this.socket.getInputStream()));
```
Comment (redcatbear) [inline src/test/java/com/exasol/LogProxy.java:52]: Readers not closed. Use try-with-resources to utilize `AutoClosable`.

File: src/test/java/com/exasol/adapter/dynamodb/ExasolTestInterface.java [hunk 2]
```diff
@@ -100,15 +113,22 @@ public void createConnection(final String name, final String to, final String us
...
         final StringBuilder statementBuilder = new StringBuilder(
                 "CREATE OR REPLACE JAVA ADAPTER SCRIPT " + ADAPTER_SCHEMA + "." + DYNAMODB_ADAPTER + " AS\n");
         final String hostIp = getTestHostIpAddress();
 
-        if (hostIp != null && !isNoDebugSystemPropertySet()) {
-            // noinspection SpellCheckingInspection
-            statementBuilder.append("  %jvmoption -agentlib:jdwp=transport=dt_socket,server=n,address=").append(hostIp)
-                    .append(":8000,suspend=y;\n");
+        if (hostIp != null) {
+            String jvmOptions = "-javaagent:/buckets/bfsdefault/default/" + JACOCO_JAR_NAME
```
Comment (redcatbear) [inline src/test/java/com/exasol/adapter/dynamodb/ExasolTestInterface.java:124]: Make final and use `?` operator or use `StringBuilder`.

File: src/test/java/com/exasol/adapter/dynamodb/ExasolTestInterface.java [hunk 3]
```diff
@@ -100,15 +113,22 @@ public void createConnection(final String name, final String to, final String us
...
+                    + "=output=tcpclient,address=" + hostIp + ",port=3002";
+            if (!isNoDebugSystemPropertySet()) {
+                // noinspection SpellCheckingInspection
+                jvmOptions += " -agentlib:jdwp=transport=dt_socket,server=n,address=" + hostIp + ":" + DEBUGGER_PORT
+                        + ",suspend=y";
+            }
+            statementBuilder.append("  %jvmoption ").append(jvmOptions).append(";\n");
```
Comment (redcatbear) [inline src/test/java/com/exasol/adapter/dynamodb/ExasolTestInterface.java:131]: Now I see where you need to modify the VS definition.

File: src/test/java/com/exasol/jacoco/JacocoServer.java [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        if (instance == null) {
+            instance = new JacocoServer();
+            new Thread(instance).start();
+        }
+    }
+
+    @Override
+    public void run() {
+        try {
+            final ServerSocket server = new ServerSocket(PORT);
```
Comment (redcatbear) [inline src/test/java/com/exasol/jacoco/JacocoServer.java:56]: Use try-with-resources.

File: src/test/java/com/exasol/LogProxy.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+/**
+ * This class opens a tcp socket and dumps everything to stdout. It is used for printing logs from the udf.
+ */
+public class LogProxy implements Runnable {
+    private final static int PORT = 3001;
+    private static LogProxy instance;
+    private final ServerSocket server;
+
+    public LogProxy() throws IOException {
+        this.server = new ServerSocket(PORT);
```
[the remark this answers is not in the document]
Comment (jakobbraun) [reply src/test/java/com/exasol/LogProxy.java:20]: Well when should I close it? in finalize?

File: src/test/java/com/exasol/adapter/dynamodb/ExasolTestInterface.java (hunk 2 above)
Comment (jakobbraun) [reply src/test/java/com/exasol/adapter/dynamodb/ExasolTestInterface.java]: I dont like to use the ? operator here. Its not an alternative. Instead something is added conditionally. I think using an if its much more readable here. How about the solution with the StringBuilder?

File: src/test/java/com/exasol/LogProxy.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (redcatbear) [reply src/test/java/com/exasol/LogProxy.java:20]: If you open it in the constructor, close it in `finalize`.

File: src/test/java/com/exasol/adapter/dynamodb/ExasolTestInterface.java (hunk 2 above)
Comment (redcatbear) [reply src/test/java/com/exasol/adapter/dynamodb/ExasolTestInterface.java]: Yes, that is also perfectly fine.

**Pull request merged**